modules/aggregation-layers/src/aggregation-layer-v9/cpu-aggregator/cpu-aggregator.ts

```diff
@@ -0,0 +1,189 @@
+import type {Aggregator, AggregationProps} from '../aggregator';
+import {_deepEqual as deepEqual, BinaryAttribute} from '@deck.gl/core';
+import {sortBins, packBinIds} from './sort';
+import {aggregateChannel} from './aggregate';
+import {VertexAccessor, evaluateVertexAccessor} from './vertex-accessor';
+
+/** Settings used to construct a new CPUAggregator */
+export type CPUAggregatorSettings = {
+  /** Size of bin IDs */
+  dimensions: number;
+  /** Accessor to map each data point to a bin ID.
+   * If dimensions=1, bin ID should be a number;
+   * If dimensions>1, bin ID should be an array with [dimensions] elements;
+   * The data point will be skipped if bin ID is null.
+   */
+  getBin: VertexAccessor<number | number[] | null, any>;
+  /** Accessor to map each data point to a weight value, defined per channel */
+  getWeight: VertexAccessor<number>[];
+};
+
+/** Options used to run CPU aggregation, can be changed at any time */
+export type CPUAggregationProps = AggregationProps & {};
+
+export type Bin = {
+  id: number | number[];
+  index: number;
+  /** list of data point indices */
+  points: number[];
+};
+
+/** An Aggregator implementation that calculates aggregation on the CPU */
+export class CPUAggregator implements Aggregator {
+  dimensions: number;
+  numChannels: number;
+
+  props: CPUAggregationProps = {
+    binOptions: {},
+    pointCount: 0,
+    operations: [],
+    attributes: {}
+  };
+
+  protected getBinId: CPUAggregatorSettings['getBin'];
```

Comment:
Nit: probably clearer to type out the callback types here, rather than referencing another type.